src/librustc_hir/lang_items.rs

```diff
@@ -47,44 +47,86 @@ macro_rules! language_item_table {
             }
         }
 
+        #[derive(Debug, Copy, Clone, Hash)]
+        pub enum LangItemRecord {
+            /// The language item was found and mapped to its `DefId`
+            /// during collection.
+            DefId(DefId),
+            /// The language item was not found during collection.
+            Missing(LangItem),
+        }
+
+        impl LangItemRecord {
+            /// Requires that the `LangItem` was bound and returns the corresponding `DefId`.
+            pub fn require(self) -> Result<DefId, LangItem> {
```

Comment:
Part of the reason to move away from `Option` is to make it hard for people to call `unwrap`. Having a function that returns a `Result` sort of defeats the purpose doesn't it? Now I just do `lang_items().sized().require().unwrap()`. I think `require` should accept a `TyCtxt` and a `span` and emit an error when it fails. If you want to do more complex things, you can match on `LangItemRecord` directly. I

Reply:
Okay yeah that’s a fair point - I’ll try again that way around although I feel like there may be some places where the code definitely take a hit

Reply:
To me, this indicates that this PR should go through the major-change proposal process as niko mentioned on the issue. MCP is relatively new, so I think we're still exploring what magnitude of change qualifies as "major". This PR is a good test of that lower bound.

Would you be interested in [opening a proposal](https://github.com/rust-lang/rfcs/blob/master/text/2904-compiler-major-change-process.md#major-change-proposals) that explains the anti-pattern this PR tries to prevent as well as a rough sketch of the API? One positive side-effect of the process is that more people will be informed about the correct use of the new API.

Reply:
A similar question-- is there a rustc-dev-guide entry on lang items? I think there **should be**, and I suspect that any such entry would be affected by this change (as it would likely have given some examples of how lang items are meant to be used).

Reply:
I don’t think we have a dev guide entry on lang items yet and I agree that we should. I’ve got a couple of exams next week but once they’re done I’ll write up the MCP, rewrite this PR and start drafting some stuff for the dev guide.

Reply:
Sounds great! Thanks for your work on this. =)

I don't think btw that the MCP has to be super detailed, I think sketching out the plan is sufficient, especially given that you can point at this PR. The main idea is to have some standard way to advertise changes people may want to be aware of.